### DISCUSSION TITLE
Build error when generating pathname nextjs 13 app dir

### DISCUSSION THREAD

**Discussion by e8ward** [vercel/next.js#45933] [Help] [answered]
Im having a problem during build. i want to dynamically get all the SEO metadata from a static data. and inject into the head.js

### DefaultTag.js
```jsx
'use client';
import { metaTags } from '@/data';
import { usePathname } from 'next/navigation';

export function generateStaticParams() {
  return metaTags.map((item) => ({
    result: item.path,
  }));
}

export default function DefaultTag() {
  let pathname = usePathname();
  const meta = metaTags.find((meta) => meta.path === pathname);
  return (
    <>
      <title>{meta.title}</title>
      <meta name='description' content={meta.description} />
      <meta name='keywords' content={meta.keywords} />
      <meta name='viewport' content='width=device-width, initial-scale=1' />
      <link rel='icon' href='/favicon.ico' />
    </>
  );
}
```

### Head.js

```jsx
import DefaultTag from './DefaultTag';

export default function Head() {
  return (
    <>
      <DefaultTag />
    </>
  );
}
```

when using dev its working fine. but during build it cannot find the data. because the pathname is not generated during build. i can't use the slug params. because its only for paths like 
```
example.com/
example.com/services
example.com/solutions
```

im wondering is there a way to statically generate the pathname. so that during build time. its can get all the possible pathnames ?
Thank You :)

**Comment by kfirfitousi** [comment]
`generateStaticParams` only works in `page.js` and `layout.js` inside dynamic segments.  

> i can't use the slug params. because its only for paths like
> ```
> example.com/
> example.com/services
> example.com/solutions
> ```

How do your paths look like?

Note that `head.js` will be deprecated in 13.2 and removed in the future.  
Instead, we get built-in support for SEO through [Metadata](https://beta.nextjs.org/docs/guides/seo), which is available now on the canary version.

**Comment by denisso** [comment]
[e8ward](https://github.com/e8ward), that's how it works

[Code repository](https://github.com/denisso/next13-app-exp/tree/metadata-for-each-page) , [sandbox](https://codesandbox.io/p/github/denisso/next13-app-exp/metadata-for-each-page?workspaceId=18dbcb21-f19f-4aca-a6ae-6859de0638c0&selection=%5B%7B%22endColumn%22%3A18%2C%22endLineNumber%22%3A4%2C%22startColumn%22%3A18%2C%22startLineNumber%22%3A4%7D%5D&file=%2FREADME.md), [deploy](https://lk2sob-3000.preview.csb.app/)

app\services\page.tsx
```tsx
import { metaTags } from "@/data";

export const metadata = {
  title: metaTags.services.title,
  description: metaTags.services.description,
  keywords: metaTags.services.keywords,
  icons: {
    icon: "/favicon.ico",
  },
};

export default async function Page (){
  return <>Service page</>
} 
```

app\solutions\page.tsx
```tsx
import { metaTags } from "@/data";

export const metadata = {
  title: metaTags.solutions.title,
  description: metaTags.solutions.description,
  keywords: metaTags.solutions.keywords,
  icons: {
    icon: "/favicon.ico",
  },
};

export default async function Page (){
  return <>Solutions page</>
} 
```

**Reply by e8ward** [reply]
Oh my... Thank you ! now its working

**Comment by e8ward** [comment, marked as answer]
After days of research and look at leerob's latest github repo this is the solution that works for me

```jsx
'use client';
import { usePathname } from 'next/navigation';
import { metaTags } from '@/data';

export async function generateMetadata() {
  const pathname = usePathname();
  const paths = metaTags.find((meta) => meta.path === pathname);
  if (!paths) {
    return;
  }
}

export default function DefaultTag() {
  const pathname = usePathname();
  const meta = metaTags.find((meta) => meta.path === pathname);
  return (
    <>
      <title>{meta.title}</title>
      <meta name='description' content={meta.description} />
      <meta name='keywords' content={meta.keywords} />
      <meta name='viewport' content='width=device-width, initial-scale=1' />
      <link rel='icon' href='/favicon.ico' />
    </>
  );
}
```

Basically you output the return to that file it will nextjs will automatically do the rest for you on how to parse and map the metadata for you
